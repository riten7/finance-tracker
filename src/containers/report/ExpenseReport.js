import React from 'react';
import { useSelector } from 'react-redux';
import { Divider } from 'semantic-ui-react';
import { DatePicker } from 'antd';
import { FilterFilled } from '@ant-design/icons';
import moment from 'moment';
import Chart from './Chart';

import './report.css'

const { RangePicker } = DatePicker;

const NoData = () => (
  <div className="transactions-form__empty">You don't have any transactions within the selected date range</div>
);

const ExpenseReport = () => {
  const { transactions } = useSelector(state => state.transactions);

  const [filteredTransactions, setFilteredTransactions] = React.useState(transactions);

  const getFormattedDate = (date) => {
    return date.format('YYYY-MM-DD');
  }

  const onDateChange = (dates) => {
    if(!dates) return setFilteredTransactions(transactions);;
    const filteredTxs = transactions.filter(tx => getFormattedDate(moment(tx.date)) >= getFormattedDate(dates[0]) && getFormattedDate(moment(tx.date)) <= getFormattedDate(dates[1]))
    setFilteredTransactions(filteredTxs);
  }

  return (
    <div className="report-container">
      <div className="report-container__picker">
      <RangePicker
        className="report_input_test"
        onChange={onDateChange}
        format={'YYYY-MM-DD'}
        suffixIcon={<FilterFilled />}/>
      </div>
      <Divider />
      {filteredTransactions.length > 0 ?
      <>
      <Chart transactions={filteredTransactions} txtype="Income" />
      <br />
      <Chart transactions={filteredTransactions} txtype="Expense" />
      </>
      : <NoData />}
    </div>
  )
};

export default ExpenseReport;
