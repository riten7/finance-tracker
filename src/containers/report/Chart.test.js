import React from "react";
import moment from "moment";
import { mount } from "enzyme";
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import ExpenseReport from "./ExpenseReport";

const data = [
  {
    "amount": 4000,
    "accountId": "6zb98plvkbr",
    "date": "2020-09-18T18:15:00.000Z",
    "note": "wrerw",
    "accountName": "dtrdtr",
    "tags": [
      "qw"
    ],
    "id": "vsmynjtll1e",
    "type": "expense"
  },
  {
    "amount": 12312,
    "accountId": "6zb98plvkbr",
    "date": "2020-09-11T18:15:00.000Z",
    "note": "asdasd",
    "accountName": "dtrdtr",
    "tags": [
      "asdasd"
    ],
    "id": "ji26szpwofq",
    "type": "expense"
  },
  {
    "amount": 4000,
    "accountId": "6zb98plvkbr",
    "date": "2020-09-15T18:15:00.000Z",
    "note": "asd",
    "accountName": "dtrdtr",
    "tags": [
      "qw"
    ],
    "id": "ul3db7uhzw",
    "type": "income"
  }
]

const mockStore = configureMockStore();

describe("Chart Component renders properly", () => {
  const setFilteredTransactions = jest.fn();
  const handleClick = jest.spyOn(React, "useState");
  handleClick.mockImplementation(size => [size, setFilteredTransactions]);

  const store = mockStore({ transactions: { transactions: data } });

  const incomeWrapper = mount(
    <Provider store={store}>
      <ExpenseReport />
    </Provider>);

  it("Component renders successfully", () => {
    expect(incomeWrapper.html()).toMatchSnapshot();
    expect(incomeWrapper.find('.k-chart')).toHaveLength(2);
  });

  it("On Click of pdf export works properly", () => {
    const onHandlePdfExport = jest.fn();
    incomeWrapper.find('.k-button').at(0).simulate('click');
    //expect(onHandlePdfExport).toBeCalled();
  });

  it("On Date Change works properly", () => {
    act(() => {
      incomeWrapper.find('RangePicker').at(0).props().onChange([moment('2020-02-02'), moment('2020-02-05')]);
    });
    expect(setFilteredTransactions).toBeTruthy();
  });
});

describe("Chart Component renders properly", () => {
  const store = mockStore({ transactions: { transactions: [] } });

  const incomeWrapper = mount(
    <Provider store={store}>
      <ExpenseReport />
    </Provider>);

  it("Component renders successfully", () => {
    expect(incomeWrapper.html()).toMatchSnapshot();
    expect(incomeWrapper.find('.transactions-form__empty')).toHaveLength(1);
  });
})