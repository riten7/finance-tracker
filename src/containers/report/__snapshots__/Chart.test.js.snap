// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart Component renders properly Component renders successfully 1`] = `"<div class=\\"report-container\\"><div class=\\"report-container__picker\\"><div class=\\"ant-picker ant-picker-range report_input_test\\"><div class=\\"ant-picker-input ant-picker-input-active\\"><input readonly=\\"\\" placeholder=\\"Start date\\" size=\\"12\\" autocomplete=\\"off\\" value=\\"\\"></div><div class=\\"ant-picker-range-separator\\"><span aria-label=\\"to\\" class=\\"ant-picker-separator\\"><span role=\\"img\\" aria-label=\\"swap-right\\" class=\\"anticon anticon-swap-right\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"swap-right\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M873.1 596.2l-164-208A32 32 0 00684 376h-64.8c-6.7 0-10.4 7.7-6.3 13l144.3 183H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h695.9c26.8 0 41.7-30.8 25.2-51.8z\\"></path></svg></span></span></div><div class=\\"ant-picker-input\\"><input readonly=\\"\\" placeholder=\\"End date\\" size=\\"12\\" autocomplete=\\"off\\" value=\\"\\"></div><div class=\\"ant-picker-active-bar\\" style=\\"left: 0px; width: 0px; position: absolute;\\"></div><span class=\\"ant-picker-suffix\\"><span role=\\"img\\" aria-label=\\"calendar\\" class=\\"anticon anticon-calendar\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"calendar\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z\\"></path></svg></span></span></div></div><div class=\\"ui divider\\"></div><button class=\\"ui button k-button\\">Export as PDF</button><div class=\\"k-chart k-widget\\" style=\\"position: relative;\\"><div class=\\"k-chart-surface\\" style=\\"position: relative;\\"><svg style=\\"width: 100%; height: 100%; overflow: hidden;\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" version=\\"1.1\\"><defs><clipPath id=\\"kdef1\\"><path d=\\"M24.5 25.5 L 594.5 25.5 594.5 376.5 24.5 376.5Z\\" stroke=\\"#000\\" stroke-linejoin=\\"miter\\" fill=\\"none\\"></path></clipPath></defs><g><path d=\\"M0 0 L 600 0 600 400 0 400Z\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M24 25 L 594 25 594 376 24 376Z\\" stroke=\\"none\\" fill=\\"#fff\\" fill-opacity=\\"0\\"></path><g><g><path d=\\"M24.5 25.5 L 594.5 25.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g><g><path d=\\"M87.5 25.5 L 87.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M151.5 25.5 L 151.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M214.5 25.5 L 214.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M277.5 25.5 L 277.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M341.5 25.5 L 341.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M404.5 25.5 L 404.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M467.5 25.5 L 467.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M531.5 25.5 L 531.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M594.5 25.5 L 594.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g><g><g><g><path d=\\"M24.5 25.5 L 24.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M20.5 25.5 L 24.5 25.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M20.5 376.5 L 24.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g></g></g><g clip-path=\\"url(#kdef1)\\"><g transform=\\"matrix(0,0,0,1,0,0)\\"><g><g><path d=\\"M24.5 130.5 L 531.5 130.5 531.5 271.5 24.5 271.5Z\\" stroke=\\"#006600\\" stroke-width=\\"1\\" stroke-opacity=\\"1\\" fill=\\"green\\" fill-opacity=\\"1\\"></path></g></g></g></g><g transform=\\"matrix(0,-1,1,0,-196,211)\\"><text style=\\"font: ;white-space:pre;\\" x=\\"10\\" y=\\"202\\" stroke=\\"none\\" fill=\\"none\\">Categories</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"16\\" y=\\"201.5\\" stroke=\\"none\\" fill=\\"none\\">qw</text></g><g><g><path d=\\"M24.5 376.5 L 594.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M24.5 376.5 L 24.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M87.5 376.5 L 87.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M151.5 376.5 L 151.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M214.5 376.5 L 214.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M277.5 376.5 L 277.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M341.5 376.5 L 341.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M404.5 376.5 L 404.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M467.5 376.5 L 467.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M531.5 376.5 L 531.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M594.5 376.5 L 594.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"24.5\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">0</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"87.833\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">500</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"151.167\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">1000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"214.5\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">1500</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"277.833\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">2000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"341.167\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">2500</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"404.5\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">3000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"467.833\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">3500</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"531.167\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">4000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"594.5\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">4500</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"310\\" y=\\"391\\" stroke=\\"none\\" fill=\\"none\\">Income</text></g></g><g><g><text style=\\"font: ;white-space:pre;\\" x=\\"300.5\\" y=\\"16\\" stroke=\\"none\\" fill=\\"none\\">Income Transaction Report</text></g></g><g></g></g></svg></div></div><div class=\\"ui divider\\"></div><br><button class=\\"ui button k-button\\">Export as PDF</button><div class=\\"k-chart k-widget\\" style=\\"position: relative;\\"><div class=\\"k-chart-surface\\" style=\\"position: relative;\\"><svg style=\\"width: 100%; height: 100%; overflow: hidden;\\" xmlns=\\"http://www.w3.org/2000/svg\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" version=\\"1.1\\"><defs><clipPath id=\\"kdef2\\"><path d=\\"M24.5 25.5 L 594.5 25.5 594.5 376.5 24.5 376.5Z\\" stroke=\\"#000\\" stroke-linejoin=\\"miter\\" fill=\\"none\\"></path></clipPath></defs><g><path d=\\"M0 0 L 600 0 600 400 0 400Z\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M24 25 L 594 25 594 376 24 376Z\\" stroke=\\"none\\" fill=\\"#fff\\" fill-opacity=\\"0\\"></path><g><g><path d=\\"M24.5 25.5 L 594.5 25.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M24.5 201.5 L 594.5 201.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g><g><path d=\\"M105.5 25.5 L 105.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M187.5 25.5 L 187.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M268.5 25.5 L 268.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M350.5 25.5 L 350.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M431.5 25.5 L 431.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M513.5 25.5 L 513.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M594.5 25.5 L 594.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g><g><g><g><path d=\\"M24.5 25.5 L 24.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M20.5 25.5 L 24.5 25.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M20.5 201.5 L 24.5 201.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M20.5 376.5 L 24.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g></g></g><g clip-path=\\"url(#kdef2)\\"><g transform=\\"matrix(0,0,0,1,0,0)\\"><g><g><path d=\\"M24.5 78.5 L 187.5 78.5 187.5 148.5 24.5 148.5Z\\" stroke=\\"#cc0000\\" stroke-width=\\"1\\" stroke-opacity=\\"1\\" fill=\\"red\\" fill-opacity=\\"1\\"></path></g></g><g><g><path d=\\"M24.5 253.5 L 525.5 253.5 525.5 323.5 24.5 323.5Z\\" stroke=\\"#cc0000\\" stroke-width=\\"1\\" stroke-opacity=\\"1\\" fill=\\"red\\" fill-opacity=\\"1\\"></path></g></g></g></g><g transform=\\"matrix(0,-1,1,0,-196,211)\\"><text style=\\"font: ;white-space:pre;\\" x=\\"10\\" y=\\"202\\" stroke=\\"none\\" fill=\\"none\\">Categories</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"16\\" y=\\"113.75\\" stroke=\\"none\\" fill=\\"none\\">qw</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"16\\" y=\\"289.25\\" stroke=\\"none\\" fill=\\"none\\">asdasd</text></g><g><g><path d=\\"M24.5 376.5 L 594.5 376.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M24.5 376.5 L 24.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M105.5 376.5 L 105.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M187.5 376.5 L 187.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M268.5 376.5 L 268.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M350.5 376.5 L 350.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M431.5 376.5 L 431.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M513.5 376.5 L 513.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path><path d=\\"M594.5 376.5 L 594.5 380.5\\" stroke=\\"none\\" fill=\\"none\\"></path></g></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"24.5\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">0</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"105.929\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">2000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"187.357\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">4000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"268.786\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">6000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"350.214\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">8000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"431.643\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">10000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"513.071\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">12000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"594.5\\" y=\\"386\\" stroke=\\"none\\" fill=\\"none\\">14000</text></g><g><text style=\\"font: ;white-space:pre;\\" x=\\"310\\" y=\\"391\\" stroke=\\"none\\" fill=\\"none\\">Expense</text></g></g><g><g><text style=\\"font: ;white-space:pre;\\" x=\\"300.5\\" y=\\"16\\" stroke=\\"none\\" fill=\\"none\\">Expense Transaction Report</text></g></g><g></g></g></svg></div></div><div class=\\"ui divider\\"></div></div>"`;

exports[`Chart Component renders properly Component renders successfully 2`] = `"<div class=\\"report-container\\"><div class=\\"report-container__picker\\"><div class=\\"ant-picker ant-picker-range report_input_test\\"><div class=\\"ant-picker-input ant-picker-input-active\\"><input readonly=\\"\\" placeholder=\\"Start date\\" size=\\"12\\" autocomplete=\\"off\\" value=\\"\\"></div><div class=\\"ant-picker-range-separator\\"><span aria-label=\\"to\\" class=\\"ant-picker-separator\\"><span role=\\"img\\" aria-label=\\"swap-right\\" class=\\"anticon anticon-swap-right\\"><svg viewBox=\\"0 0 1024 1024\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"swap-right\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M873.1 596.2l-164-208A32 32 0 00684 376h-64.8c-6.7 0-10.4 7.7-6.3 13l144.3 183H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h695.9c26.8 0 41.7-30.8 25.2-51.8z\\"></path></svg></span></span></div><div class=\\"ant-picker-input\\"><input readonly=\\"\\" placeholder=\\"End date\\" size=\\"12\\" autocomplete=\\"off\\" value=\\"\\"></div><div class=\\"ant-picker-active-bar\\" style=\\"left: 0px; width: 0px; position: absolute;\\"></div><span class=\\"ant-picker-suffix\\"><span role=\\"img\\" aria-label=\\"calendar\\" class=\\"anticon anticon-calendar\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"calendar\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z\\"></path></svg></span></span></div></div><div class=\\"ui divider\\"></div><div class=\\"transactions-form__empty\\">You don't have any transactions within the selected date range</div></div>"`;
